File: fix_viva.js
const fs = require('fs');

function walkDir(dir) {
    let results = [];
    const list = fs.readdirSync(dir);
    list.forEach(function(file) {
        file = dir + '/' + file;
        const stat = fs.statSync(file);
        if (stat && stat.isDirectory()) { 
            results = results.concat(walkDir(file));
        } else { 
            if(file.endsWith('.js')) {
                results.push(file);
            }
        }
    });
    return results;
}

const files = walkDir('experiments');

files.forEach(file => {
    let content = fs.readFileSync(file, 'utf8');
    
    // Non-greedy match for anything between onclick="window.VIVA_SYSTEM.open( and )"
    const newContent = content.replace(
        /onclick="window\.VIVA_SYSTEM\.open\([\s\S]*?\}\)"/g,
        'onclick="window.VIVA_SYSTEM.open(${JSON.stringify(VIVA_QUESTIONS).replace(/&/g, \'&amp;\').replace(/\\"/g, \'&quot;\')})"'
    );
    
    if (newContent !== content) {
        fs.writeFileSync(file, newContent, 'utf8');
        console.log('Fixed ' + file);
    }
});

File: fix_viva.py
import os

def fix_viva_buttons():
    for root, dirs, files in os.walk('experiments'):
        for file in files:
            if file.endswith('.js'):
                path = os.path.join(root, file)
                with open(path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                # The exact string that caused the issue:
                broken_str = r'onclick="window.VIVA_SYSTEM.open(${JSON.stringify(VIVA_QUESTIONS).replace(/&/g, \'&amp;\').replace(/\"/g, \'&quot;\')})"'
                
                # The correct string without backslashes before single quotes and no escaped backslash before double quote
                fixed_str = r'onclick="window.VIVA_SYSTEM.open(${JSON.stringify(VIVA_QUESTIONS).replace(/&/g, \'&amp;\').replace(/\"/g, \'&quot;\')})"'.replace(r"\'", "'").replace(r'\"', '"')
                # Wait, fixed_str should just be:
                fixed_str = r"""onclick="window.VIVA_SYSTEM.open(${JSON.stringify(VIVA_QUESTIONS).replace(/&/g, '&amp;').replace(/"/g, '&quot;')})\""""
                # Actually, I should just explicitly define it without raw strings if it's confusing:
                fixed_str = 'onclick="window.VIVA_SYSTEM.open(${JSON.stringify(VIVA_QUESTIONS).replace(/&/g, \'&amp;\').replace(/\\"/g, \'&quot;\')})"'
                
                # Let's cleanly define them:
                broken_str = "onclick=\"window.VIVA_SYSTEM.open(${JSON.stringify(VIVA_QUESTIONS).replace(/&/g, \\'&amp;\\').replace(/\\\"/g, \\'&quot;\\')})\""
                fixed_str = "onclick=\"window.VIVA_SYSTEM.open(${JSON.stringify(VIVA_QUESTIONS).replace(/&/g, '&amp;').replace(/\\\"/g, '&quot;')})\""
                
                if broken_str in content:
                    new_content = content.replace(broken_str, fixed_str)
                    with open(path, 'w', encoding='utf-8') as f:
                        f.write(new_content)
                    print(f'Fixed {path}')

if __name__ == "__main__":
    fix_viva_buttons()
